### PR TITLE
[#5316] Upgrade to Django 4.2

### LINKED ISSUE #5316: Upgrade to Django 4.2

### PR DESCRIPTION
**Pull request by pkdash**
Resolves #5316 

### Pull Request Checklist: 
- [ ] Positive Test Case Written by Dev

- [x] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [x] Peer Code review and approval

### Positive Test Case
1. [Enter positive test case here]

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -4,6 +4,16 @@ FROM hydroshare/hs_docker_base:0bc6b1d
 RUN sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && \
     locale-gen
 
+RUN pip install django==4.2.*
+RUN pip install djangorestframework==3.14.*
+RUN pip install django-haystack==3.2.*
+RUN pip install drf-yasg==1.21.*
+RUN pip install django-robots==6.1
+RUN pip install django-autocomplete-light==3.9.*
```
Comment (devincowan) [inline Dockerfile:12]: https://django-autocomplete-light.readthedocs.io/en/master/index.html

> Django 4.2 support starting DAL 3.10, support for earlier Django versions in DAL < 3.10

**Comment by pkdash** [comment]
@devincowan I have now updated it with develop. You can work on autocomplete issue now. I still have to do bit of cleanup which I will do after your work.

**Comment by devincowan** [comment]
@pkdash I'm hoping to find time to pick this back up this week (finally)--thanks for your patience!
Do you want to attempt to resolve conflicts before proceeding or should I?
If my memory serves, I was going to fix (or replace?) a dropdown element that was broken due to auto-complete UI...

**Comment by devincowan** [comment]
Also, entirely fine if this needs to wait until after the CSV work. I just thought I might have a couple hours this week to look at this again

**Comment by pkdash** [comment]
@devincowan Let me know when you would have time to work on fixing the autocomplete that is currently broken due to the upgrade and I will resolve these code conflicts accordingly. Thanks.

File: Dockerfile (hunk 1 above)
Comment (pkdash) [reply Dockerfile]: There is now DAL 3.11

**Comment by devincowan** [comment]
@pkdash I believe I have fixed all of the autocomplete instances...but it would be good to have a second set of eyes to make sure.

**Comment by pkdash** [comment]
@devincowan I can do some testing of autocomplete later today.

**Comment by pkdash** [comment]
@devincowan I did test the autocomplete and it seems to work. Thanks for fixing the autocomplete UI related issues.

**Comment by devincowan** [comment]
Sounds good, I will hope to do a full review next week!

**Review by devincowan — DISMISSED** [review]
looks good...wow, a lot of work!
I just pushed one change to fix an issue that I introduced.

Other than that, I think we will need to rely heavily on QA for this one.

**Comment by devincowan** [comment]
yeah I think that's advisable.
My guess is that we do CSV and 4.2 as a release (after QA) since they are ready to go

File: Dockerfile-multistage-node [hunk 2]
```diff
@@ -36,6 +36,18 @@ COPY --from=node-build /hydroshare /hydroshare
 RUN sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && \
     locale-gen
 
+RUN pip install psycopg==3.1.*
```
Comment (pkdash) [inline Dockerfile-multistage-node:39]: Do we need to remove the install commands from here?
https://github.com/hydroshare/hydroshare/pull/5352/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R8

Comment (devincowan) [reply Dockerfile-multistage-node:39]: no they are intentional there until we get them into the base image

**Review by devincowan — APPROVED** [review]

**Comment by devincowan** [comment]
@pkdash I approved, ready to merge

**Pull request merged**